tests/vanilla/types.test.tsx

```diff
@@ -67,3 +67,12 @@ it('type utils should work', () => {
   }
   expect(Component).toBeDefined()
 })
+
+it('WritableAtom Result type should be covariant', () => {
+  function Component() {
+    const writableAtom = atom(null, () => null)
```

Comment:
This test doesn't represent your original reproduction. Can you do something similar to your repro in the PR description?

Reply:
I mean in this case, we could do:
```suggestion
    const writableAtom = atom(null, () => null as null | number)
```

I think we need the actual case in the test:
```ts
export const createWritableAtom = (): WritableAtom<
  null,
  [],
  number | null
> =>
  atom(
    () => null,
    (get, set, someArg: string) => {
      return null;
    },
  );
```